### PR TITLE
add capabilities endpoint

### PR DESCRIPTION
**Pull request by SimoneDutto**
## Description

Add capabilities endpoint and tests

> fly by improve get identity error not found

Fixes _JIRA/GitHub issue number_

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

File: internal/rebac_admin/capabilities_test.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	req, err := http.NewRequest("GET", url, nil)
+	c.Assert(err, qt.IsNil)
+	resp, err := http.DefaultClient.Do(req)
+	c.Assert(err, qt.IsNil)
+	defer resp.Body.Close()
+	c.Assert(resp.StatusCode, qt.Equals, 404)
+
+	// test endpoints in capabilities are found
+	for _, cap := range rebac_admin.Capabilities {
+		for _, m := range cap.Methods {
```
Comment (kian99) [inline internal/rebac_admin/capabilities_test.go:39]: Is there any way to test the inverse as well? I.e. verify that we included all the endpoints that we are capable of in the list.

**Review by kian99 — DISMISSED** [review]

File: internal/rebac_admin/identities.go [hunk 2]
```diff
@@ -190,7 +194,10 @@ func (s *identitiesService) GetIdentityEntitlements(ctx context.Context, identit
 	}
 	objUser, err := s.jimm.FetchIdentity(ctx, identityId)
 	if err != nil {
-		return nil, v1.NewNotFoundError(fmt.Sprintf("User with id %s not found", identityId))
+		if errors.ErrorCode(err) == errors.CodeNotFound {
+			return nil, v1.NewNotFoundError("failed to find identity")
```
Comment (babakks) [inline internal/rebac_admin/identities.go]: Can you please include the ID of the identity in the error message?

File: internal/rebac_admin/capabilities.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		Endpoint: "/capabilities",
+		Methods: []resources.CapabilityMethods{
+			"GET",
+		},
+	},
+	{
+		Endpoint: "/identities",
+		Methods: []resources.CapabilityMethods{
+			"GET",
+			"POST",
```
Comment (babakks) [inline internal/rebac_admin/capabilities.go:40]: Remove `POST`, because we do not support it.

File: internal/rebac_admin/capabilities.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+		Endpoint: "/identities/{id}/groups",
+		Methods: []resources.CapabilityMethods{
+			"GET",
+			"PATCH",
+		},
+	},
+	{
+		Endpoint: "/identities/{id}/roles",
+		Methods:  []resources.CapabilityMethods{},
+	},
```
Comment (babakks) [inline internal/rebac_admin/capabilities.go]: Remove this entry.

File: internal/rebac_admin/capabilities.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		Endpoint: "/groups/{id}/identities",
+		Methods: []resources.CapabilityMethods{
+			"GET",
+			"PATCH",
+		},
+	},
+	{
+		Endpoint: "/groups/{id}/roles",
+		Methods:  []resources.CapabilityMethods{},
+	},
```
Comment (babakks) [inline internal/rebac_admin/capabilities.go]: Remove this entry.

File: internal/rebac_admin/capabilities.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	{
+		Endpoint: "/roles/{id}",
+		Methods:  []resources.CapabilityMethods{},
+	},
+	{
+		Endpoint: "/roles/{id}/entitlements",
+		Methods:  []resources.CapabilityMethods{},
+	},
+	{
```
Comment (babakks) [inline internal/rebac_admin/capabilities.go]: Remove all `/roles` related entries.

File: internal/rebac_admin/capabilities_test.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	defer testServer.Close()
+
+	// test not found endpoint
+	url := fmt.Sprintf("%s/v1%s", testServer.URL, "not-found")
+	req, err := http.NewRequest("GET", url, nil)
+	c.Assert(err, qt.IsNil)
+	resp, err := http.DefaultClient.Do(req)
+	c.Assert(err, qt.IsNil)
+	defer resp.Body.Close()
+	c.Assert(resp.StatusCode, qt.Equals, 404)
```
Comment (babakks) [inline internal/rebac_admin/capabilities_test.go]: Why do we need this part?

File: internal/rebac_admin/capabilities_test.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	for _, cap := range rebac_admin.Capabilities {
+		for _, m := range cap.Methods {
+			c.Run(fmt.Sprintf("%s %s", m, cap.Endpoint), func(c *qt.C) {
+				url := fmt.Sprintf("%s/v1%s", testServer.URL, cap.Endpoint)
+				req, err := http.NewRequest(string(m), url, nil)
+				c.Assert(err, qt.IsNil)
+				resp, err := http.DefaultClient.Do(req)
+				c.Assert(err, qt.IsNil)
+				defer resp.Body.Close()
+				c.Assert(resp.StatusCode, qt.Not(qt.Equals), 404)
```
Comment (babakks) [inline internal/rebac_admin/capabilities_test.go]: The library never returns `404` on an unimplemented endpoint (instead returns `501 Unimplemented`).

Thanks to this PR, I just noticed the library panics if you invoke an entirely unimplemented realm (like `/roles`).

File: internal/rebac_admin/capabilities.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
+// Copyright 2024 Canonical.
+package rebac_admin
```
Comment (alesstimec) [inline internal/rebac_admin/capabilities.go:3]: new line between copyright and package, please

File: internal/rebac_admin/capabilities.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+import (
+	"context"
+
+	"github.com/canonical/rebac-admin-ui-handlers/v1/interfaces"
+	"github.com/canonical/rebac-admin-ui-handlers/v1/resources"
+)
+
+// CapabilitiesService implements the `CapabilitiesService` interface.
```
Comment (alesstimec) [inline internal/rebac_admin/capabilities.go]: lower case

File: internal/rebac_admin/capabilities_test.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2024 Canonical.
+package rebac_admin_test
```
Comment (alesstimec) [inline internal/rebac_admin/capabilities_test.go:3]: new line between copyright and package

File: internal/rebac_admin/capabilities_test.go (hunk 1 above)
Comment (alesstimec) [reply internal/rebac_admin/capabilities_test.go:39]: json encode rebac_admin.Capabilities and assert that the json we get is the same as the encoded capabilities?

**Review by alesstimec — DISMISSED** [review]
LGTM with Babak's comments addressed

File: internal/rebac_admin/capabilities_test.go (hunk 8 above)
Comment (SimoneDutto) [reply internal/rebac_admin/capabilities_test.go]: yes, i noticed the same thing. What's the expected behaviour in the future? A 501?

Comment (SimoneDutto) [reply internal/rebac_admin/capabilities_test.go]: 404 is still possible for entirely wrong endpoint, look at the test before this loop

File: internal/rebac_admin/capabilities_test.go (hunk 1 above)
Comment (SimoneDutto) [reply internal/rebac_admin/capabilities_test.go:39]: i think in this way we are just the library is json encoding well.

@kian99 is asking for a way to test that all endpoints under /rebac are declared by this capabilities endpoint and there are no endpoints slipping away

File: internal/rebac_admin/capabilities_test.go (hunk 8 above)
Comment (babakks) [reply internal/rebac_admin/capabilities_test.go]: You're right. Can you please split the assertion against `501` and `404` and explain them? I mean, to clarify that we assert against `404` to make sure the path is not something invalid, and against `501` to make sure the endpoint is not left unimplmeneted.

File: internal/rebac_admin/capabilities_test.go (hunk 1 above)
Comment (babakks) [reply internal/rebac_admin/capabilities_test.go:39]: Doing that can be possible if the library provides a method that returns a *full* capability data. That way you know about all endpoints and can iterate over them and assert if they're implemented. But, I think this is something that the library should provide, and JIMM should not be aware of such raw details of the OpenAPI spec.

**Review by kian99 — APPROVED** [review]

**Review by babakks — APPROVED** [review]

**Review by ale8k — APPROVED** [review]

**Pull request merged**